### DISCUSSION TITLE
Timing problems with creating a batch of notes. Can't get tp.hooks.on_all_templates_executed(callback_function: () => any working

### DISCUSSION THREAD

**Discussion by Elkiehirvi** [SilentVoid13/Templater#1493] [Help] [answered]
I'm having a bit of a problem getting my recurring calendar event template working. The following correctly creates the events when executed from my regular calendar event template as I intended, and it populates the metadata, but the template referred in creating the event _"Calendar event recurring instance"_ creates a bit of text on the event note that pulls from the metadata - which is null when the template resolves.
```
<%* 
    let numb = await tp.system.prompt("Days between recurrences");
    var category = await tp.system.prompt("f-c category:");
    var timeS = await tp.system.prompt("time start: hh:mm");
    var timeE = await tp.system.prompt("time end: hh:mm");
    var display = await tp.system.prompt("fc-display: ");
    var place = await tp.system.prompt("Location of event: ");
    var root = tp.file.title;
    let snick = root.split(/\s+/);
    var name = snick[0];
    const dates = tp.user.enumerateBetweenDates(tp.frontmatter["recur-start"], tp.frontmatter["recur-end"], numb);
    dates.forEach(createInstance)

    async function createInstance(date) {
        await tp.file.create_new(tp.file.find_tfile("Calendar event recurring instance"), `${date}` + " " + name, false, "10 Journal/10 Calendar" )
        var file = await tp.file.find_tfile(`${date}` + " " + name);
        await app.fileManager.processFrontMatter(file, (frontmatter) => {
        	frontmatter["fc-date"] = `${date}`;
        	frontmatter["fc-category"] = category;
        	frontmatter["time-start"] = timeS;
        	frontmatter["time-end"] = timeE;
        	frontmatter["fc-display-name"] = display;
        	frontmatter["location"] = place;
            });
    };
%>
```
 The "_recurring instance_" template works if I put the metadata in manually before resolving, but when creating the note via **tp.file.create_new()** the order of templates resolving is wrong, and the metadata is not populated before the new note tries to resolve.

As I understand it **tp.hooks.on_all_templates_executed(callback_function: () => any)** should work for exactly this problem, but when I use it with my metadata referencing text generation snippet in the _Calendar event recurring instance_ - template it resolves into a blank note. I also tried using it within the loop in the above script that creates the notes, but it also failed - partly because I couldn't figure how to output text into the new note as tR += didn't work, which I kinf of figured.
I'm having a bit of a problem getting my recurring calendar event template working. The following correctly creates the events when executed from my regular calendar event template as I intended, and it populates the metadata, but the template referred in creating the event _"Calendar event recurring instance"_ creates a bit of text on the event note that pulls from the metadata - which is null when the template resolves.

The "_recurring instance_" template works if I put the metadata in manually before resolving, but when creating the note via **tp.file.create_new()** the order of templates resolving is wrong, and the metadata is not populated before the new note tries to resolve.

As I understand it **tp.hooks.on_all_templates_executed(callback_function: () => any)** should work for exactly this problem, but when I use it with my metadata referencing text generation snippet in the _Calendar event recurring instance_ - template it resolves into a blank note. I also tried using it within the loop in the above script that creates the notes, but it also failed - partly because I couldn't figure how to output text into the new note as tR += didn't work, which I kind of figured.

I've tried with wait times and hooks.on_all_templates and both: 

```
<%*
tp.hooks.on_all_templates_executed(() => {
    setTimeout(() => {
        tR += "# " + tp.file.title + "\n"
        let varDate = moment(tp.frontmatter["fc-date"])
        let myDate = varDate.format("dddd MMMM YYYY")
        
        let text = "The " + "**" + tp.frontmatter["fc-display-name"] + "**" + " takes place on " + "**" + `${myDate}` + "**" + " during " + "**" + tp.frontmatter["time-start"] + " - " + tp.frontmatter["time-end"] + "**" + " at " + "**" + tp.frontmatter["location"] + "**";
        tR += text
    }, 200) // the reason for the timeout is to let the template complete first
})
-%>
```

Anyway, all help with this would be appreciated, whether on how to get **tp.hooks.on_all_templates_executed(callback_function: () => any)** working, or how to output text from the foreach() loop into a new note. Or just a better way to accomplish what I am trying to do.

Anyway, all help with this would be appreciated, whether on how to get **tp.hooks.on_all_templates_executed(callback_function: () => any)** working, or how to output text from the foreach() loop into a new note. Or just a better way to accomplish what I am trying to do.

**Comment by Zachatoo** [comment, marked as answer]
> the order of templates resolving is wrong

This because you are creating the files concurrently (see #1405 for another example of this). I would recommend creating the files sequentially. This will likely resolve most of the issues you're seeing.

```js
// 🔴 creating files concurrently
dates.forEach(createInstance);

// ✅ creating files sequentially
for (const date in dates) {
  await createInstance(date);
}
```

As far as the hooks module goes, I suspect you should be able to implement it like this in your provided template, if you still need it.

```diff
// ...
async function createInstance(date) {
  // ...
+ tp.hooks.on_all_templates_executed(async () => {
    await app.fileManager.processFrontMatter(file, (frontmatter) => {
      // process frontmatter...
    });
+ });
}
```

**Reply by Elkiehirvi** [reply]
Thank you so much for a clear and thoughtful answer. It seems my lack of experience with JS in general strikes again. The problem seems to indeed have been the concurrent creation of the files, and everything works now as one would expect. I'm marking this answered, and thank you again. This was extremely helpful.
